timelock-corruption-detection/src/test/java/com/palantir/history/HistoryLoaderAndTransformerTest.java

```diff
@@ -0,0 +1,183 @@
+        LogsForNamespaceAndUseCase logsForNamespaceAndUseCase = paxosHistory.get(0);
+
+        assertThat(logsForNamespaceAndUseCase.getNamespaceAndUseCase().namespace()).isEqualTo(CLIENT);
+        assertThat(logsForNamespaceAndUseCase.getNamespaceAndUseCase().useCase()).isEqualTo(USE_CASE);
+        assertThat(logsForNamespaceAndUseCase.getLogs().size()).isEqualTo(100 - lastVerified);
+
+        PaxosLogWithAcceptedAndLearnedValues singleLog = logsForNamespaceAndUseCase.getLogs().get(0);
+        assertThat(singleLog.getAcceptedState()).isPresent();
+        assertThat(singleLog.getPaxosValue()).isNotPresent();
+    }
+
+    @Test
+    public void canHandleHistoryWithOnlyLearnerLogs() {
+        IntStream.range(0, 100).forEach(i -> {
+            writeValueForLogAndRound(learnerLog, i + 1);
+        });
+
+        int lastVerified = 52;
+        List<HistoryQuery> historyQueries = ImmutableList.of(HistoryQuery.of(
+                ImmutableNamespaceAndUseCase.of(CLIENT, USE_CASE), lastVerified));
+
+        List<LogsForNamespaceAndUseCase> paxosHistory
+                = HistoryLoaderAndTransformer.getLogsForHistoryQueries(history, historyQueries);
+
+        assertThat(paxosHistory.size()).isEqualTo(1);
+        LogsForNamespaceAndUseCase logsForNamespaceAndUseCase = paxosHistory.get(0);
+
+        assertThat(logsForNamespaceAndUseCase.getNamespaceAndUseCase().namespace()).isEqualTo(CLIENT);
+        assertThat(logsForNamespaceAndUseCase.getNamespaceAndUseCase().useCase()).isEqualTo(USE_CASE);
+        assertThat(logsForNamespaceAndUseCase.getLogs().size()).isEqualTo(100 - lastVerified);
+
+        PaxosLogWithAcceptedAndLearnedValues singleLog = logsForNamespaceAndUseCase.getLogs().get(0);
+        assertThat(singleLog.getAcceptedState()).isNotPresent();
+        assertThat(singleLog.getPaxosValue()).isPresent();
+    }
+
+    @Test
+    public void canHandleHistoryWithNoLogs() {
+        int lastVerified = 102;
+        List<HistoryQuery> historyQueries = ImmutableList.of(HistoryQuery.of(
+                ImmutableNamespaceAndUseCase.of(CLIENT, USE_CASE), lastVerified));
+
+        List<LogsForNamespaceAndUseCase> paxosHistory
+                = HistoryLoaderAndTransformer.getLogsForHistoryQueries(history, historyQueries);
+
+        assertThat(paxosHistory.size()).isEqualTo(1);
+        LogsForNamespaceAndUseCase logsForNamespaceAndUseCase = paxosHistory.get(0);
+
+        assertThat(logsForNamespaceAndUseCase.getNamespaceAndUseCase().namespace()).isEqualTo(CLIENT);
+        assertThat(logsForNamespaceAndUseCase.getNamespaceAndUseCase().useCase()).isEqualTo(USE_CASE);
+        assertThat(logsForNamespaceAndUseCase.getLogs().size()).isEqualTo(0);
+    }
+
+    private void writeToLogs(int range) {
+        IntStream.range(0, range).forEach(i -> {
+            writeAcceptorStateForLogAndRound(acceptorLog, i + 1);
+            writeValueForLogAndRound(learnerLog, i + 1);
+        });
+    }
```

Comment:
This is pretty solid. I think you might want to add a test for a discontinuity (i.e. logs only exist from 45 - 100, and lastVerified was 3 or something like that), but other than that there's good coverage here.